src/data-model-providers/codegen/CodegenDataModelProvider.cpp

```diff
@@ -448,74 +263,43 @@ DataModel::MetadataList<DataModel::ServerClusterEntry> CodegenDataModelProvider:
-        if (err != CHIP_NO_ERROR)
-        {
-#if CHIP_ERROR_LOGGING && CHIP_CONFIG_DATA_MODEL_EXTRA_LOGGING
-            ChipLogError(AppServer, "Failed to append server cluster entry: %" CHIP_ERROR_FORMAT, err.Format());
-#endif
-            break;
-        }
+        ReturnErrorOnFailure(builder.Append(ServerClusterEntryFrom(endpointId, *cluster)));
     }
 
-    return result;
+    return CHIP_NO_ERROR;
 }
 
-DataModel::MetadataList<DataModel::AttributeEntry> CodegenDataModelProvider::Attributes(const ConcreteClusterPath & path)
+CHIP_ERROR CodegenDataModelProvider::Attributes(const ConcreteClusterPath & path,
+                                                DataModel::ListBuilder<DataModel::AttributeEntry> & builder)
 {
     const EmberAfCluster * cluster = FindServerCluster(path);
 
-    DataModel::MetadataList<DataModel::AttributeEntry> result;
-
-    VerifyOrReturnValue(cluster != nullptr, result);
-    VerifyOrReturnValue(cluster->attributeCount > 0, result);
-    VerifyOrReturnValue(cluster->attributes != nullptr, result);
+    VerifyOrReturnValue(cluster != nullptr, CHIP_ERROR_NOT_FOUND);
+    VerifyOrReturnValue(cluster->attributeCount > 0, CHIP_NO_ERROR);
+    VerifyOrReturnValue(cluster->attributes != nullptr, CHIP_NO_ERROR);
 
-    CHIP_ERROR err = result.Reserve(cluster->attributeCount);
-    if (err != CHIP_NO_ERROR)
-    {
-#if CHIP_ERROR_LOGGING && CHIP_CONFIG_DATA_MODEL_EXTRA_LOGGING
-        ChipLogError(AppServer, "Failed to reserve space for attributes: %" CHIP_ERROR_FORMAT, err.Format());
-#endif
-        return {};
-    }
+    // TODO: if ember would encode data in AttributeEntry form, we could reference things directly
+    ReturnErrorOnFailure(builder.EnsureAppendCapacity(cluster->attributeCount));
 
     Span<const EmberAfAttributeMetadata> attributeSpan(cluster->attributes, cluster->attributeCount);
 
     for (auto & attribute : attributeSpan)
     {
-        err = result.Append(AttributeEntryFrom(path, attribute));
-        if (err != CHIP_NO_ERROR)
-        {
-#if CHIP_ERROR_LOGGING && CHIP_CONFIG_DATA_MODEL_EXTRA_LOGGING
-            ChipLogError(AppServer, "Failed to append attribute entry: %" CHIP_ERROR_FORMAT, err.Format());
-#endif
-            break;
-        }
+        ReturnErrorOnFailure(builder.Append(AttributeEntryFrom(path, attribute)));
     }
 
-    return result;
+    return CHIP_NO_ERROR;
 }
 
-DataModel::MetadataList<ClusterId> CodegenDataModelProvider::ClientClusters(EndpointId endpointId)
+CHIP_ERROR CodegenDataModelProvider::ClientClusters(EndpointId endpointId, DataModel::ListBuilder<ClusterId> & builder)
 {
     const EmberAfEndpointType * endpoint = emberAfFindEndpointType(endpointId);
 
-    DataModel::MetadataList<ClusterId> result;
+    VerifyOrReturnValue(endpoint != nullptr, CHIP_ERROR_NOT_FOUND);
+    VerifyOrReturnValue(endpoint->clusterCount > 0, CHIP_NO_ERROR);
+    VerifyOrReturnValue(endpoint->cluster != nullptr, CHIP_NO_ERROR);
```

Comment:
Can combine condition (and everywhere similar?)

Reply:
I find not combining is clearer on readability and breakpoints (if applicable) . What is the reason you prefer combining?

If this would be an actual error I would strongly prefer separation so I know exactly what part failed ... but for CHIP_NO_ERROR I find it just readable ...

Reply:
Checked size differences if I combine in QPG and only debug info changes. I think we can keep them separate for readability:

![image](https://github.com/user-attachments/assets/9392290f-5a5a-42fe-9182-cc24a8379e82)